README.md
# NAME

Catmandu::FileStore - Namespace for packages that can make files persistent

# STATUS

[![Build Status](https://travis-ci.org/LibreCat/Catmandu-FileStore.svg?branch=master)](https://travis-ci.org/LibreCat/Catmandu-FileStore)
[![Coverage](https://coveralls.io/repos/LibreCat/Catmandu-FileStore/badge.svg?branch=master)](https://coveralls.io/r/LibreCat/Catmandu-FileStore)

# SYNOPSIS

    # From the command line

    # Export a list of all file containers
    $ catmandu export File::Simple --root t/data to YAML

    # Export a list of all files in container '1234'
    $ catmandu export File::Simple --root t/data --bag 1234 to YAML

    # Add a file to the container '1234'
    $ catmandu stream /tmp/myfile.txt to File::Simple --root t/data --bag 1234 --id myfile.txt

    # Download the file 'myfile.txt' from the container '1234'
    $ catmandu stream File::Simple --root t/data --bag 1234 --id myfile.txt to /tmp/output.txt

    # Delete the file 'myfile.txt' from the container '1234'
    $ catmandu delete File::Simple --root t/data --bag 1234 --id myfile.txt

    # From Perl
    use Catmandu;

    my $store = Catmandu->store('File::Simple' , root => 't/data');

    # List all containers
    $store->index->each(sub {
        my $container = shift;

        print "%s\n" , $container->{_id};
    });

    # Add a new container
    $store->index->add({ _id => '1234' });

    # Get the container
    my $files = $store->index->files('1234');

    # Add a file to the container
    $files->upload(IO::File->new('<foobar.txt'), 'foobar.txt');

    my $file = $files->get('foobar.txt');

    # Stream the contents of a file
    $files->stream(IO::File->new('>foobar.txt'), $file);

    # Delete a file
    $files->delete('foobar.txt');

    # Delete a container
    $store->index->delete('1234');

# DESCRIPTION

Each [Catmandu::FileStore](https://metacpan.org/pod/Catmandu::FileStore) is a [Catmandu::Store](https://metacpan.org/pod/Catmandu::Store) and inherits all its methods,

A [Catmandu::FileStore](https://metacpan.org/pod/Catmandu::FileStore) is package to store and retrieve binary content in
an filesystem, memory or a network. A `Catmandu::FileStore` contains one or more
`Catmandu::FileBag` which is a kind of folder.

Each `Catmandu::FileBag` contains one or more files.

One special `Catmandu::FileBag` is the `index` and contains the listing
of all `Catmandu::FileBag` in the `Catmandu::FileStore`.

# CONFIGURATION

- index\_bag

    The name of the index bag to use when no bag name is give. The index bag is a
    bag containing a listing of all `Catmandu::FileBag`-s in the Store.

        my $index = $store->index;

        $index->each(sub {
            my $bag = shift;

            printf "%s\n" , $bag->{_id};
        });

- index\_class

    The default class implementation to use for an index of `Catmandu::FileBag`-s.
    By default this is the `Catmandu::FileStore` implementation with '::Index' added.

# METHODS

## bag($name)

Create or retieve a bag with name `$name`. Returns a [Catmandu::FileBag](https://metacpan.org/pod/Catmandu::FileBag).

## index

Returns the index  [Catmandu::FileBag](https://metacpan.org/pod/Catmandu::FileBag) for the [Catmandu::FileStore](https://metacpan.org/pod/Catmandu::FileStore).

    my $index = $store->index;

    # Add a new file container
    $index->add({ _id => '1234'});

    # Anf use it...
    my $container = $store->bag('1234');

    $container->upload(IO::File->new('data.txt') , 'data.txt');

## log

Return the current logger. Can be used when creating your own Stores.

E.g.

    package Catmandu::Store::Hash;

    ...

    sub generator {
        my ($self) = @_;

        $self->log->debug("generating record");
        ...
    }

See also: [Catmandu](https://metacpan.org/pod/Catmandu) for activating the logger in your main code.

# AUTHOR

Patrick Hochstenbach, `<patrick.hochstenbach at ugent.be>`

Nicolas Steenlant, `<nicolas.steenlant at ugent.be>`

Nicolas Franck, `<nicolas.franck at ugent.be>`

# SEE ALSO

[Catmandu::Store::File::Simple](https://metacpan.org/pod/Catmandu::Store::File::Simple),
[Catmandu::Store::File::Memory](https://metacpan.org/pod/Catmandu::Store::File::Memory),
[Catmandu::FileBag](https://metacpan.org/pod/Catmandu::FileBag)

# LICENSE AND COPYRIGHT

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See [http://dev.perl.org/licenses/](http://dev.perl.org/licenses/) for more information.


Changes
Revision history for Catmandu-FileStore

{{$NEXT}}

1.16  2019-12-16 15:33:10 CET
    - Fixing tests for numbered directories
    - Adding a more robust helper to stream file data

1.151  2019-10-11 07:43:46 CEST
    - Adding missing test dependencies [manwar]

1.15  2019-10-08 09:55:11 CEST
    - Adding missing dependencies [manwar]

1.14  2019-08-20 09:56:05 CEST
    - Removing Data::UUID dependency

1.13  2018-10-18 14:09:14 CEST
    - Fixing EAGAIN bug

1.12  2018-08-29 14:22:03 CEST
    - Replacing writes with syswrites and better checks on number of bytes written

1.11  2018-06-28 11:30:58 CEST
    - Fixing Windows tests

1.10  2018-06-25 13:37:20 CEST
    - initial version


lib/Catmandu/FileStore.pm
package Catmandu::FileStore;

our $VERSION = '1.16';

use Catmandu::Sane;
use Moo::Role;
use Catmandu::Util;
use namespace::clean;

with 'Catmandu::Store';

has index_bag   => (is => 'ro', default => sub {'index'},);
has index_class => (is => 'ro', default => sub {ref($_[0]) . '::Index'},);
has index       => (is => 'lazy');

sub _build_default_bag {
    $_[0]->index_bag;
}

sub _build_index {
    my ($self) = @_;
    my $name = $self->index_bag;
    my $inst;

    try {
        my $opts         = {store => $self, name => $name};
        my $default_opts = $self->default_options;
        my $bag_opts     = $self->bag_options->{$name} //= {};
        $opts = {%$default_opts, %$bag_opts, %$opts};

        my $pkg        = Catmandu::Util::require_package($self->index_class);
        my $index_name = $self->index_bag;

        my $default_plugins = $self->default_plugins;
        my $plugins = delete($opts->{plugins}) // [];

        if (@$default_plugins || @$plugins) {
            $pkg = $pkg->with_plugins(@$default_plugins, @$plugins);
        }

        $inst = $pkg->new(%$opts);
    }
    catch {
        $self->log->warn(
            "no instance of " . $self->index_class . " created : $_");
    };

    $inst;
}

sub bag {
    my $self       = shift;
    my $name       = shift // $self->index_bag;
    my $index_name = $self->index_bag;

    # Return the index when requested
    if ($name eq $index_name) {
        $self->index;
    }

    # Otherwise load the container for files
    elsif ($self->index->exists($name)) {
        my $opts         = {store => $self, name => $name};
        my $default_opts = $self->default_options;
        my $bag_opts     = $self->bag_options->{$name} //= {};
        $opts = {%$default_opts, %$bag_opts, %$opts};
        my $pkg = Catmandu::Util::require_package(delete($opts->{class})
                // $self->bag_class);

        my $default_plugins = $self->default_plugins;
        my $plugins = delete($opts->{plugins}) // [];

        if (@$default_plugins || @$plugins) {
            $pkg = $pkg->with_plugins(@$default_plugins, @$plugins);
        }

        $pkg->new(%$opts);
    }
    else {
        return undef;
    }
}

1;

__END__

=pod

=head1 NAME

Catmandu::FileStore - Namespace for packages that can make files persistent

=begin markdown

# STATUS

[![Build Status](https://travis-ci.org/LibreCat/Catmandu-FileStore.svg?branch=master)](https://travis-ci.org/LibreCat/Catmandu-FileStore)
[![Coverage](https://coveralls.io/repos/LibreCat/Catmandu-FileStore/badge.svg?branch=master)](https://coveralls.io/r/LibreCat/Catmandu-FileStore)

=end markdown

=head1 SYNOPSIS

    # From the command line

    # Export a list of all file containers
    $ catmandu export File::Simple --root t/data to YAML

    # Export a list of all files in container '1234'
    $ catmandu export File::Simple --root t/data --bag 1234 to YAML

    # Add a file to the container '1234'
    $ catmandu stream /tmp/myfile.txt to File::Simple --root t/data --bag 1234 --id myfile.txt

    # Download the file 'myfile.txt' from the container '1234'
    $ catmandu stream File::Simple --root t/data --bag 1234 --id myfile.txt to /tmp/output.txt

    # Delete the file 'myfile.txt' from the container '1234'
    $ catmandu delete File::Simple --root t/data --bag 1234 --id myfile.txt

    # From Perl
    use Catmandu;

    my $store = Catmandu->store('File::Simple' , root => 't/data');

    # List all containers
    $store->index->each(sub {
        my $container = shift;

        print "%s\n" , $container->{_id};
    });

    # Add a new container
    $store->index->add({ _id => '1234' });

    # Get the container
    my $files = $store->index->files('1234');

    # Add a file to the container
    $files->upload(IO::File->new('<foobar.txt'), 'foobar.txt');

    my $file = $files->get('foobar.txt');

    # Stream the contents of a file
    $files->stream(IO::File->new('>foobar.txt'), $file);

    # Delete a file
    $files->delete('foobar.txt');

    # Delete a container
    $store->index->delete('1234');

=head1 DESCRIPTION

Each L<Catmandu::FileStore> is a L<Catmandu::Store> and inherits all its methods,

A L<Catmandu::FileStore> is package to store and retrieve binary content in
an filesystem, memory or a network. A C<Catmandu::FileStore> contains one or more
C<Catmandu::FileBag> which is a kind of folder.

Each C<Catmandu::FileBag> contains one or more files.

One special C<Catmandu::FileBag> is the C<index> and contains the listing
of all C<Catmandu::FileBag> in the C<Catmandu::FileStore>.

=head1 CONFIGURATION

=over

=item index_bag

The name of the index bag to use when no bag name is give. The index bag is a
bag containing a listing of all C<Catmandu::FileBag>-s in the Store.

    my $index = $store->index;

    $index->each(sub {
        my $bag = shift;

        printf "%s\n" , $bag->{_id};
    });

=item index_class

The default class implementation to use for an index of C<Catmandu::FileBag>-s.
By default this is the C<Catmandu::FileStore> implementation with '::Index' added.

=back

=head1 METHODS

=head2 bag($name)

Create or retieve a bag with name C<$name>. Returns a L<Catmandu::FileBag>.

=head2 index

Returns the index  L<Catmandu::FileBag> for the L<Catmandu::FileStore>.

  my $index = $store->index;

  # Add a new file container
  $index->add({ _id => '1234'});

  # Anf use it...
  my $container = $store->bag('1234');

  $container->upload(IO::File->new('data.txt') , 'data.txt');

=head2 log

Return the current logger. Can be used when creating your own Stores.

E.g.

    package Catmandu::Store::Hash;

    ...

    sub generator {
        my ($self) = @_;

        $self->log->debug("generating record");
        ...
    }

See also: L<Catmandu> for activating the logger in your main code.

=head1 AUTHOR

Patrick Hochstenbach, C<< <patrick.hochstenbach at ugent.be> >>

Nicolas Steenlant, C<< <nicolas.steenlant at ugent.be> >>

Nicolas Franck, C<< <nicolas.franck at ugent.be> >>

=head1 SEE ALSO

L<Catmandu::Store::File::Simple>,
L<Catmandu::Store::File::Memory>,
L<Catmandu::FileBag>

=head1 LICENSE AND COPYRIGHT

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut
